File: Go-Egypt/src/app/services/attraction.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http'; // 1. ضيفنا HttpParams
import { Observable } from 'rxjs';

// --- Interfaces (DTOs) ---
export interface PaginatedResponse<T> {
  pageIndex: number;
  pageSize: number;
  count: number;
  data: T[];
}

export interface HomeAttractions {
  id: number;
  name: string;
  mainPhotoPath: string;
  location: string;
  openingTime: string;
  closingTime: string;
}

export interface CardAttractions {
  id: number;
  name: string;
  location: string;
  openingTime: string;
  closingTime: string;
  overview: string;
  mainPhotoPath: string;
  category: string;
}

export interface KeyFact {
  headline: string;
  description: string;
}

export interface AttractionDetails {
  id: number;
  name: string;
  location: string;
  overview: string;
  openingTime: string;
  closingTime: string;
  mainPhotoPath: string;
  category: string;
  gallery: string[];
  keyFacts: KeyFact[];
}

@Injectable({
  providedIn: 'root'
})
export class AttractionService { 

  private apiUrl = 'https://localhost:7212/api/Attraction'; 

  constructor(private http: HttpClient) { }

  getAttractions(page: number, size: number, category?: string): Observable<PaginatedResponse<CardAttractions>> {
    
    let params = new HttpParams()
      .set('PageIndex', page.toString())
      .set('PageSize', size.toString());

    if (category) {
      params = params.set('Category', category);
    }

    return this.http.get<PaginatedResponse<CardAttractions>>(
      `${this.apiUrl}/GetAllCardAttractions`, 
      { params }
    );
  }

  getHomeAttractions(): Observable<HomeAttractions[]> {
    return this.http.get<HomeAttractions[]>(
      `${this.apiUrl}/GetHomeAttractions`
    );
  }

  getAttractionById(id: number): Observable<AttractionDetails> {
    return this.http.get<AttractionDetails>(
      `${this.apiUrl}/GetAttractionById/${id}`
    );
  }
}

File: backend/GoEgypt Solution/Domain/Models/Attraction.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Domain.Models
{
    public class Attraction
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public string Location { get; set; }
        public string Overview { get; set; }
        public TimeOnly OpeningTime { get; set; }
        public TimeOnly ClosingTime { get; set; }
        public List<KeyFact> KeyFacts { get; set; }
        public string MainPhotoPath { get; set; }
        public string Category { get; set; }
        public List<string> Gallery { get; set; }
        // Base price per adult per day; defaults to 120 if not provided in seed/data
        public decimal Price { get; set; } = 120m;
    }

    public class KeyFact
    {
        public string Headline { get; set; }
        public string Description { get; set; }
    }
}
